native/src/main/java/io/xlibb/pipe/Pipe.java

```diff
@@ -106,20 +114,25 @@ protected void asyncProduce(Callback callback, Object event, long timeout) {
         }
     }
 
-    protected void asyncConsume(Callback callback, long timeout) {
+    protected void asyncConsume(Callback callback, long timeout, Type type) {
         if (this.queue == null) {
             callback.onSuccess(null);
             return;
         }
         synchronized (this.consumeLock) {
-            if (this.queueSize.get() == 0) {
+            if (this.queueSize.get() == 0 || queue.isEmpty()) {
```

Comment:
Do we need both these conditions ? I think `queue.isEmpty()` is the only required thing here.

Reply:
The `queue` here is a `ConcurrentLinkedQueue`, which is used for concurrent operations. And since the `isEmpty()` and `size()` APIs are not better to use in concurrent applications, an atomic integer is used to keep track of elements in the queue. But on rare occasions, under heavy load, despite using an atomic variable the queue can get empty. To fix this `queue.isEmpty()` is used as an additional check.